---
title: "Step06.nsites.Rmd"
author: "Vu lab"
date: "2023-06-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# nsites
```{r}
library(tidyverse)
library(rstatix)
library(ggpubr)
library(ggplot2)
library(limma)
library(edgeR)


```

## nsites_CDS
### scatter plot

```{r}
library(tidyverse)
load(file = "output/Step03.C2U_QC.rdata")
load(file = "output/Step02.TranscriptID_Geneid_Genename.rdata")


# count_gene_nsites function
count_gene_nsites <- function(sample, prefix) {
  sample_gene_nsites <- sample %>%
    filter(biotype %in% c("CDS"))%>%
    dplyr::select(Geneid, start) %>%
    distinct() %>% 
    group_by(Geneid) %>%
    count() %>%
    rename(!!paste0(prefix, "_nsites") := n) 
  
  return(sample_gene_nsites)
}



SCR_Rep1_gene_nsites <- count_gene_nsites(SCR_Rep1,"SCR_Rep1")
SCR_Rep2_gene_nsites <- count_gene_nsites(SCR_Rep2,"SCR_Rep2")
SCR_Rep3_gene_nsites <- count_gene_nsites(SCR_Rep3,"SCR_Rep3")

shRNA33_Rep1_gene_nsites <- count_gene_nsites(shRNA33_Rep1,"shRNA33_Rep1")
shRNA33_Rep2_gene_nsites <- count_gene_nsites(shRNA33_Rep2,"shRNA33_Rep2")
shRNA33_Rep3_gene_nsites <- count_gene_nsites(shRNA33_Rep3,"shRNA33_Rep3")

shRNA37_Rep1_gene_nsites <- count_gene_nsites(shRNA37_Rep1,"shRNA37_Rep1")
shRNA37_Rep2_gene_nsites <- count_gene_nsites(shRNA37_Rep2,"shRNA37_Rep2")
shRNA37_Rep3_gene_nsites <- count_gene_nsites(shRNA37_Rep3,"shRNA37_Rep3")


# Merge the data frames using full_join()
gene_nsites <- full_join(SCR_Rep1_gene_nsites, SCR_Rep2_gene_nsites, by = "Geneid") %>%
  full_join(SCR_Rep3_gene_nsites, by = "Geneid") %>%
  full_join(shRNA33_Rep1_gene_nsites, by = "Geneid") %>%
  full_join(shRNA33_Rep2_gene_nsites, by = "Geneid") %>%
  full_join(shRNA33_Rep3_gene_nsites, by = "Geneid") %>%
  full_join(shRNA37_Rep1_gene_nsites, by = "Geneid") %>%
  full_join(shRNA37_Rep2_gene_nsites, by = "Geneid") %>%
  full_join(shRNA37_Rep3_gene_nsites, by = "Geneid")

# Replace NA values with 0
gene_nsites <- gene_nsites %>%
  mutate_all(~ ifelse(is.na(.), 0, .)) 

class(gene_nsites) # [1] "grouped_df" "tbl_df"     "tbl"        "data.frame"

# ungroup data frame
gene_nsites <- ungroup(gene_nsites)





```





#### SCR

```{r}
# Load required libraries
library(ggplot2)
library(dplyr)
library(ggpmisc)



p1 <- ggplot(gene_nsites, aes(x = SCR_Rep1_nsites, y = SCR_Rep2_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "control", x = "nsites rep1", y = "nsites rep2") +
  theme_classic() +
  theme(aspect.ratio = 1)


# SCR_Rep1_Rep3

p2 <- ggplot(gene_nsites, aes(x = SCR_Rep1_nsites, y = SCR_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "control", x = "nsites rep1", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


# SCR_Rep2_Rep3

p3 <- ggplot(gene_nsites, aes(x = SCR_Rep2_nsites, y = SCR_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "control", x = "nsites rep2", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


library(gridExtra)

# combine the plots
control_nsites_scattar <- grid.arrange(p1, p2, p3, nrow = 1)



ggsave("output/Step06.control_nsites_CDS_scattar.pdf", plot = control_nsites_scattar, device = "pdf", width = 8, height = 6)
```

#### shRNA33


```{r}
# Load required libraries
library(ggplot2)
library(dplyr)
library(ggpmisc)



p1 <- ggplot(gene_nsites, aes(x = shRNA33_Rep1_nsites, y = shRNA33_Rep2_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-33", x = "nsites rep1", y = "nsites rep2") +
  theme_classic() +
  theme(aspect.ratio = 1)


# shRNA33_Rep1_Rep3

p2 <- ggplot(gene_nsites, aes(x = shRNA33_Rep1_nsites, y = shRNA33_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-33", x = "nsites rep1", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


# shRNA33_Rep2_Rep3

p3 <- ggplot(gene_nsites, aes(x = shRNA33_Rep2_nsites, y = shRNA33_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-33", x = "nsites rep2", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


library(gridExtra)

# combine the plots
KD33_nsites_scattar <- grid.arrange(p1, p2, p3, nrow = 1)

ggsave("output/Step06.KD33_nsites_CDS_scattar.pdf", plot = KD33_nsites_scattar, device = "pdf", width = 8, height = 6)

```

#### shRNA37


```{r}
# Load required libraries
library(ggplot2)
library(dplyr)
library(ggpmisc)



p1 <- ggplot(gene_nsites, aes(x = shRNA37_Rep1_nsites, y = shRNA37_Rep2_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-37", x = "nsites rep1", y = "nsites rep2") +
  theme_classic() +
  theme(aspect.ratio = 1)


# shRNA37_Rep1_Rep3

p2 <- ggplot(gene_nsites, aes(x = shRNA37_Rep1_nsites, y = shRNA37_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-37", x = "nsites rep1", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


# shRNA37_Rep2_Rep3

p3 <- ggplot(gene_nsites, aes(x = shRNA37_Rep2_nsites, y = shRNA37_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-37", x = "nsites rep2", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


library(gridExtra)

# combine the plots
KD37_nsites_scattar <- grid.arrange(p1, p2, p3, nrow = 1)

ggsave("output/Step06.KD37_nsites_CDS_scattar.pdf", plot = KD37_nsites_scattar, device = "pdf", width = 8, height = 6)
```













### Quartile_BoxPlot 
#### SCR, shRNA33, shRNA37, nsites bins cut by SCR nsites value

```{r}

gene_nsites_bin <- gene_nsites %>%  
  rename_all(~ gsub("_nsites", "", .)) %>%
  mutate_all(~ ifelse(is.na(.), 0, .)) %>%
  arrange(desc((SCR_Rep1 + SCR_Rep2 + SCR_Rep3)/3))%>% 
  left_join(Geneid_Genename, by = "Geneid") %>%
  relocate(Genename, .after = Geneid)

# Calculate the number of rows in each bin 
rows_per_bin <- nrow(gene_nsites_bin) / 4

# Print the number of genes in each quartile
print(paste("Each quartile genes number:", rows_per_bin))


# Add a new "bin" column based on the row index
gene_nsites_bin <- gene_nsites_bin %>%
  mutate(bin = as.integer((row_number() - 1) / rows_per_bin) + 1)

# Reshape the data to long format
long_data <- gene_nsites_bin %>%
  gather(key = "sample_Rep", value = "nsites", SCR_Rep1, SCR_Rep2, SCR_Rep3, shRNA33_Rep1, shRNA33_Rep2, shRNA33_Rep3, shRNA37_Rep1, shRNA37_Rep2, shRNA37_Rep3) %>%
  mutate(sample_group = case_when(
    grepl("SCR", sample_Rep) ~ "SCR",
    grepl("shRNA33", sample_Rep) ~ "shRNA33",
    grepl("shRNA37", sample_Rep) ~ "shRNA37"
  ))

# Perform pairwise Wilcoxon rank-sum tests
pairwise_comparisons <- long_data %>%
  group_by(bin) %>%
  pairwise_wilcox_test(nsites ~ sample_group, p.adjust.method = "bonferroni")

# Filter the significant comparisons (e.g., p-value < 0.05)
nsites_3_significant_comparisons <- pairwise_comparisons %>%
  filter(p.adj < 0.05)

print(nsites_3_significant_comparisons)

# Calculate the average EPKM value for each sample group in each bin
average_nsites <- long_data %>%
  group_by(bin, sample_group) %>%
  summarize(mean_nsites = mean(nsites),
            median_nsites = median(nsites),
            .groups = "drop")

# Print the average EPKM values
print(average_nsites)


# save as excel
library(openxlsx)

# Create a new workbook
wb <- createWorkbook()

# Add the data frames as sheets to the workbook
## sheet1
addWorksheet(wb, "gene_nsites")
writeData(wb, "gene_nsites", gene_nsites)

## sheet2
addWorksheet(wb, "nsites_bins")
writeData(wb, "nsites_bins", long_data)

## sheet3
addWorksheet(wb, "3groups_significant")
writeData(wb, "3groups_significant", nsites_3_significant_comparisons)

## sheet4
addWorksheet(wb, "average_nsites")
writeData(wb, "average_nsites", average_nsites)

# Save the workbook to a file
saveWorkbook(wb, "output/Step06.nsites_CDS_Quartile_binBynsites_3groups.xlsx", overwrite = TRUE)





# Generate the box plot
colors <- c("#C1C0C0", "#76D7FF", "#1796FF")


nsites_Quartile_binBynsites_3groups <- ggplot(long_data, aes(x = as.factor(bin), y = log2(nsites + 1), fill = sample_group)) +
  geom_boxplot(color = "grey50", outlier.size = 0.5) +
  labs(x = "Quartile", y = "log2(nsites + 1)", title = "nsites (CDS) rank by control nsites value") +
  scale_fill_manual(values = colors, 
                    labels = c("control", "KD-33", "KD-37"),
                    name = "") +
  theme_classic() +
  theme(legend.position = c(0.82, 0.95),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.background = element_rect(fill = "transparent", color = NA),
        legend.text = element_text(size = 9),
        legend.title = element_text(size = 9),
        legend.key.size = unit(0.3, "cm"),
        aspect.ratio = 2/1)



ggsave("output/Step06.nsites_CDS_Quartile_binBynsites_3groups.pdf", plot = nsites_Quartile_binBynsites_3groups, device = "pdf", width = 8, height = 6)






```

#### SCR, shRNA, nsites bins cut by SCR nsites value


```{r}
# Reshape the data to long format
long_data <- gene_nsites_bin %>%
  gather(key = "sample_Rep", value = "nsites", SCR_Rep1, SCR_Rep2, SCR_Rep3, shRNA33_Rep1, shRNA33_Rep2, shRNA33_Rep3, shRNA37_Rep1, shRNA37_Rep2, shRNA37_Rep3) %>%
  mutate(sample_group = case_when(
    grepl("SCR", sample_Rep) ~ "SCR",
    grepl("shRNA33", sample_Rep) ~ "shRNA33",
    grepl("shRNA37", sample_Rep) ~ "shRNA37"
  )) %>%
  mutate(combined_group = case_when(
    sample_group == "SCR" ~ "SCR",
    sample_group %in% c("shRNA33", "shRNA37") ~ "shRNA"
  ))

# Perform pairwise Wilcoxon rank-sum tests
pairwise_comparisons <- long_data %>%
  group_by(bin) %>%
  pairwise_wilcox_test(nsites ~ combined_group, p.adjust.method = "bonferroni", pool.sd = FALSE)

# Filter the significant comparisons (e.g., p-value < 0.05)
nsites_2_significant_comparisons <- pairwise_comparisons %>%
  filter(p.adj < 0.05)

print(nsites_2_significant_comparisons)



# Calculate the average EPKM value for each sample group in each bin
average_nsites <- long_data %>%
  group_by(bin, combined_group) %>%
  summarize(mean_nsites = mean(nsites),
            median_nsites = median(nsites),
            .groups = "drop")

# Print the average EPKM values
print(average_nsites)


# save as excel
library(openxlsx)

# Create a new workbook
wb <- createWorkbook()

# Add the data frames as sheets to the workbook
## sheet1
addWorksheet(wb, "gene_nsites")
writeData(wb, "gene_nsites", gene_nsites)

## sheet2
addWorksheet(wb, "nsites_bins")
writeData(wb, "nsites_bins", long_data)

## sheet3
addWorksheet(wb, "2groups_significant")
writeData(wb, "2groups_significant", nsites_2_significant_comparisons)

## sheet4
addWorksheet(wb, "average_nsites")
writeData(wb, "average_nsites", average_nsites)

# Save the workbook to a file
saveWorkbook(wb, "output/Step06.nsites_CDS_Quartile_binBynsites_2groups.xlsx", overwrite = TRUE)





# Generate the box plot
colors <- c("#C1C0C0", "#1796FF")


nsites_Quartile_binBynsites_2groups <- ggplot(long_data, aes(x = as.factor(bin), y = log2(nsites + 1), fill = combined_group)) +
  geom_boxplot(color = "grey50", outlier.size = 0.5) +
  labs(x = "Quartile", y = "log2(nsites + 1)", title = "nsites (CDS) rank by control nsites value") +
  scale_fill_manual(values = colors, 
                    labels = c("control", "KD"),
                    name = "") +
  theme_classic() +
  theme(legend.position = c(0.82, 0.95),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.background = element_rect(fill = "transparent", color = NA),
        legend.text = element_text(size = 9),
        legend.title = element_text(size = 9),
        legend.key.size = unit(0.3, "cm"),
        aspect.ratio = 2/1)


ggsave("output/Step06.nsites_CDS_Quartile_binBynsites_2groups.pdf", plot = nsites_Quartile_binBynsites_2groups, device = "pdf", width = 8, height = 6)



save(gene_nsites_bin, file = "output/Step06.nsites_CDS_Quartile_binBynsites_2groups.rdata")

```




### Differential results
```{r}

gene_nsites <- gene_nsites %>%  
  rename_all(~ gsub("_nsites", "", .)) %>%
  mutate_all(~ ifelse(is.na(.), 0, .)) %>%
  column_to_rownames(var = "Geneid")



# Load necessary libraries
library(limma)
library(edgeR)
library(ggplot2)

# Subset the gene_nsites data frame to include only rows where the row sum is greater than 0
gene_nsites <- gene_nsites[which(rowSums(gene_nsites) > 0),]

# Display the dimensions of the resulting data frame
dim(gene_nsites)

# Create a factor variable condition with three levels
condition <- factor(paste(rep(c("SCR", "shRNA33", "shRNA37"), each = 3)))

coldata <- data.frame(row.names = colnames(gene_nsites), condition)

# Create a DGEList object from the normalized gene expression data
dge <- DGEList(counts = gene_nsites)

# Apply the TMM normalization method using edgeR's calcNormFactors function
dge <- calcNormFactors(dge)

# Convert the normalized data to log2-counts per million (logCPM) using limma's voom function
v <- voom(dge, normalize.method = "none")

# Set up the design matrix for your experimental conditions
design <- model.matrix(~0 + condition)
colnames(design) <- levels(condition)

# Fit the linear model and perform differential expression analysis
fit <- lmFit(v, design)
contrast.matrix <- makeContrasts(shRNA33_vs_SCR = shRNA33 - SCR, shRNA37_vs_SCR = shRNA37 - SCR, shRNA37_vs_shRNA33 = shRNA37 - shRNA33, levels = design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)

# Extract the top differentially expressed genes for each contrast
res_shRNA33_SCR <- topTable(fit2, number = Inf, coef = "shRNA33_vs_SCR")
res_shRNA37_SCR <- topTable(fit2, number = Inf, coef = "shRNA37_vs_SCR")
res_shRNA37_shRNA33 <- topTable(fit2, number = Inf, coef = "shRNA37_vs_shRNA33")

# Perform PCA on the logCPM values
pca <- prcomp(t(v$E), scale = TRUE)

# Create a data frame with the first two principal components and the sample conditions
pca_df <- data.frame(Sample = rownames(pca$x),
                     PC1 = pca$x[, 1],
                     PC2 = pca$x[, 2],
                     Condition = coldata$condition)

# Plot the PCA with ggplot2
ggplot(pca_df, aes(x = PC1, y = PC2, color = Condition)) +
  geom_point(size = 3) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  labs(title = "PCA plot", x = "PC1", y = "PC2") +
  scale_color_discrete(name = "Condition")








# Extract the replicate information from the sample names
pca_df$Replicate <- gsub(".*(Rep\\d+).*", "\\1", pca_df$Sample)

# Plot the PCA with ggplot2 and add sample labels
ggplot(pca_df, aes(x = PC1, y = PC2, color = Condition)) +
  geom_point(size = 3) +
  geom_text(aes(label = Replicate), hjust = 0.3, vjust = 1.3, size = 3) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  labs(title = "PCA plot", x = "PC1", y = "PC2") +
  scale_color_discrete(name = "Condition")




















# create a factor variable condition with three levels
condition2 <- factor(c("SCR", "SCR", "SCR", "shRNA", "shRNA", "shRNA", "shRNA", "shRNA", "shRNA"))


coldata <- data.frame(row.names = colnames(gene_nsites), condition2)

# Create a DGEList object from the normalized gene expression data
dge <- DGEList(counts = gene_nsites)

# Apply the TMM normalization method using edgeR's calcNormFactors function
dge <- calcNormFactors(dge)

# Convert the normalized data to log2-counts per million (logCPM) using limma's voom function
v <- voom(dge, normalize.method = "none")

# Set up the design matrix for your experimental condition2s
design <- model.matrix(~0 + condition2)
colnames(design) <- levels(condition2)

# Fit the linear model and perform differential expression analysis
fit <- lmFit(v, design)
contrast.matrix <- makeContrasts(shRNA_vs_SCR = shRNA - SCR, levels = design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)

# Extract the top differentially expressed genes for each contrast
res_shRNA_SCR <- topTable(fit2, number = Inf, coef = "shRNA_vs_SCR")


# Perform PCA on the logCPM values
pca <- prcomp(t(v$E), scale = TRUE)

# Create a data frame with the first two principal components and the sample condition2s
pca_df <- data.frame(Sample = rownames(pca$x),
                     PC1 = pca$x[, 1],
                     PC2 = pca$x[, 2],
                     condition2 = coldata$condition2)

# Plot the PCA with ggplot2
ggplot(pca_df, aes(x = PC1, y = PC2, color = condition2)) +
  geom_point(size = 3) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  labs(title = "PCA plot", x = "PC1", y = "PC2") +
  scale_color_discrete(name = "condition2")













#### results#############
library(tidyverse)
# res_shRNA33_SCR
res_shRNA33_SCR <- as.data.frame(res_shRNA33_SCR)
res_shRNA33_SCR <- res_shRNA33_SCR %>% 
  rownames_to_column(var = "Geneid") %>%
  left_join(Geneid_Genename, by="Geneid") %>%
  dplyr::select(Geneid, Genename, log2FoldChange = logFC, pvalue = P.Value, padj = adj.P.Val) %>% 
  mutate(direction = if_else(
    padj > 0.05, 'ns', if_else(
      abs(log2FoldChange) == 0, 'ns', if_else(
        log2FoldChange > 0, 'up', 'down')
    ))) %>%  
  arrange(desc(abs(log2FoldChange))) 

## group count
group_by(res_shRNA33_SCR, direction) %>%
  summarise(count=n())


# res_shRNA37_SCR
res_shRNA37_SCR <- as.data.frame(res_shRNA37_SCR)
res_shRNA37_SCR <- res_shRNA37_SCR %>% 
  rownames_to_column(var = "Geneid") %>%
  left_join(Geneid_Genename, by="Geneid") %>%
  dplyr::select(Geneid, Genename, log2FoldChange = logFC, pvalue = P.Value, padj = adj.P.Val) %>% 
  mutate(direction = if_else(
    padj > 0.05, 'ns', if_else(
      abs(log2FoldChange) == 0, 'ns', if_else(
        log2FoldChange > 0, 'up', 'down')
    ))) %>%  
  arrange(desc(abs(log2FoldChange)))   

## group count
group_by(res_shRNA37_SCR, direction) %>%
  summarise(count=n())


# res_shRNA37_shRNA33
res_shRNA37_shRNA33 <- as.data.frame(res_shRNA37_shRNA33)
res_shRNA37_shRNA33 <- res_shRNA37_shRNA33 %>% 
  rownames_to_column(var = "Geneid") %>%
  left_join(Geneid_Genename, by="Geneid") %>%
  dplyr::select(Geneid, Genename, log2FoldChange = logFC, pvalue = P.Value, padj = adj.P.Val) %>% 
  mutate(direction = if_else(
    padj > 0.05, 'ns', if_else(
      abs(log2FoldChange) == 0, 'ns', if_else(
        log2FoldChange > 0, 'up', 'down')
    ))) %>%  
  arrange(desc(abs(log2FoldChange))) 

## group count
group_by(res_shRNA37_shRNA33, direction) %>%
  summarise(count=n())


# res_shRNA_SCR
res_shRNA_SCR <- as.data.frame(res_shRNA_SCR)
res_shRNA_SCR <- res_shRNA_SCR %>% 
  rownames_to_column(var = "Geneid") %>%
  left_join(Geneid_Genename, by="Geneid") %>%
  dplyr::select(Geneid, Genename, log2FoldChange = logFC, pvalue = P.Value, padj = adj.P.Val) %>% 
  mutate(direction = if_else(
    padj > 0.05, 'ns', if_else(
      abs(log2FoldChange) == 0, 'ns', if_else(
        log2FoldChange > 0, 'up', 'down')
    ))) %>%  
  arrange(desc(abs(log2FoldChange))) 

## group count
group_by(res_shRNA_SCR, direction) %>%
  summarise(count=n())














######       save
# save as excel
library(openxlsx)

# Create a new workbook
wb <- createWorkbook()

# Add the data frames as sheets to the workbook
## res_shRNA33_SCR
addWorksheet(wb, "res_shRNA33_SCR")
writeData(wb, "res_shRNA33_SCR", res_shRNA33_SCR)

## res_shRNA37_SCR
addWorksheet(wb, "res_shRNA37_SCR")
writeData(wb, "res_shRNA37_SCR", res_shRNA37_SCR)

## res_shRNA37_shRNA33
addWorksheet(wb, "res_shRNA37_shRNA33")
writeData(wb, "res_shRNA37_shRNA33", res_shRNA37_shRNA33)

## res_shRNA_SCR
addWorksheet(wb, "res_shRNA_SCR")
writeData(wb, "res_shRNA_SCR", res_shRNA_SCR)

# Save the workbook to a file
saveWorkbook(wb, "output/Step06.nsites_CDS_DifferentialAnalysis.xlsx", overwrite = TRUE)






res_shRNA33_SCR_nsites <- res_shRNA33_SCR  

res_shRNA37_SCR_nsites <- res_shRNA33_SCR 

res_shRNA37_shRNA33_nsites <- res_shRNA33_SCR 

res_shRNA_SCR_nsites <- res_shRNA33_SCR 


save(res_shRNA33_SCR_nsites,res_shRNA33_SCR_nsites,res_shRNA37_SCR_nsites,res_shRNA37_shRNA33_nsites,res_shRNA_SCR_nsites, file = "output/Step06.nsites_CDS_DifferentialAnalysis.rdata")


```




## nsites_CDS_3UTR 

### scatter plot

```{r}
library(tidyverse)
load(file = "output/Step03.C2U_QC.rdata")
load(file = "output/Step02.TranscriptID_Geneid_Genename.rdata")


# count_gene_nsites function
count_gene_nsites <- function(sample, prefix) {
  sample_gene_nsites <- sample %>%
    filter(biotype %in% c("CDS", "three_prime_utr"))%>%
    dplyr::select(Geneid, start) %>%
    distinct() %>% 
    group_by(Geneid) %>%
    count() %>%
    rename(!!paste0(prefix, "_nsites") := n) 
  
  return(sample_gene_nsites)
}



SCR_Rep1_gene_nsites <- count_gene_nsites(SCR_Rep1,"SCR_Rep1")
SCR_Rep2_gene_nsites <- count_gene_nsites(SCR_Rep2,"SCR_Rep2")
SCR_Rep3_gene_nsites <- count_gene_nsites(SCR_Rep3,"SCR_Rep3")

shRNA33_Rep1_gene_nsites <- count_gene_nsites(shRNA33_Rep1,"shRNA33_Rep1")
shRNA33_Rep2_gene_nsites <- count_gene_nsites(shRNA33_Rep2,"shRNA33_Rep2")
shRNA33_Rep3_gene_nsites <- count_gene_nsites(shRNA33_Rep3,"shRNA33_Rep3")

shRNA37_Rep1_gene_nsites <- count_gene_nsites(shRNA37_Rep1,"shRNA37_Rep1")
shRNA37_Rep2_gene_nsites <- count_gene_nsites(shRNA37_Rep2,"shRNA37_Rep2")
shRNA37_Rep3_gene_nsites <- count_gene_nsites(shRNA37_Rep3,"shRNA37_Rep3")


# Merge the data frames using full_join()
gene_nsites <- full_join(SCR_Rep1_gene_nsites, SCR_Rep2_gene_nsites, by = "Geneid") %>%
  full_join(SCR_Rep3_gene_nsites, by = "Geneid") %>%
  full_join(shRNA33_Rep1_gene_nsites, by = "Geneid") %>%
  full_join(shRNA33_Rep2_gene_nsites, by = "Geneid") %>%
  full_join(shRNA33_Rep3_gene_nsites, by = "Geneid") %>%
  full_join(shRNA37_Rep1_gene_nsites, by = "Geneid") %>%
  full_join(shRNA37_Rep2_gene_nsites, by = "Geneid") %>%
  full_join(shRNA37_Rep3_gene_nsites, by = "Geneid")

# Replace NA values with 0
gene_nsites <- gene_nsites %>%
  mutate_all(~ ifelse(is.na(.), 0, .)) 

class(gene_nsites) # [1] "grouped_df" "tbl_df"     "tbl"        "data.frame"

# ungroup data frame
gene_nsites <- ungroup(gene_nsites)





```





#### SCR

```{r}
# Load required libraries
library(ggplot2)
library(dplyr)
library(ggpmisc)



p1 <- ggplot(gene_nsites, aes(x = SCR_Rep1_nsites, y = SCR_Rep2_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "control", x = "nsites rep1", y = "nsites rep2") +
  theme_classic() +
  theme(aspect.ratio = 1)


# SCR_Rep1_Rep3

p2 <- ggplot(gene_nsites, aes(x = SCR_Rep1_nsites, y = SCR_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "control", x = "nsites rep1", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


# SCR_Rep2_Rep3

p3 <- ggplot(gene_nsites, aes(x = SCR_Rep2_nsites, y = SCR_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "control", x = "nsites rep2", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


library(gridExtra)

# combine the plots
control_nsites_scattar <- grid.arrange(p1, p2, p3, nrow = 1)



ggsave("output/Step06.control_nsites_CDS_3UTR_scattar.pdf", plot = control_nsites_scattar, device = "pdf", width = 8, height = 6)
```

#### shRNA33


```{r}
# Load required libraries
library(ggplot2)
library(dplyr)
library(ggpmisc)



p1 <- ggplot(gene_nsites, aes(x = shRNA33_Rep1_nsites, y = shRNA33_Rep2_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-33", x = "nsites rep1", y = "nsites rep2") +
  theme_classic() +
  theme(aspect.ratio = 1)


# shRNA33_Rep1_Rep3

p2 <- ggplot(gene_nsites, aes(x = shRNA33_Rep1_nsites, y = shRNA33_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-33", x = "nsites rep1", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


# shRNA33_Rep2_Rep3

p3 <- ggplot(gene_nsites, aes(x = shRNA33_Rep2_nsites, y = shRNA33_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-33", x = "nsites rep2", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


library(gridExtra)

# combine the plots
KD33_nsites_scattar <- grid.arrange(p1, p2, p3, nrow = 1)

ggsave("output/Step06.KD33_nsites_CDS_3UTR_scattar.pdf", plot = KD33_nsites_scattar, device = "pdf", width = 8, height = 6)

```

#### shRNA37


```{r}
# Load required libraries
library(ggplot2)
library(dplyr)
library(ggpmisc)



p1 <- ggplot(gene_nsites, aes(x = shRNA37_Rep1_nsites, y = shRNA37_Rep2_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-37", x = "nsites rep1", y = "nsites rep2") +
  theme_classic() +
  theme(aspect.ratio = 1)


# shRNA37_Rep1_Rep3

p2 <- ggplot(gene_nsites, aes(x = shRNA37_Rep1_nsites, y = shRNA37_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-37", x = "nsites rep1", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


# shRNA37_Rep2_Rep3

p3 <- ggplot(gene_nsites, aes(x = shRNA37_Rep2_nsites, y = shRNA37_Rep3_nsites)) +
  geom_point(size = 2, alpha = 0.5, color = "#2277B5") +
  stat_poly_eq(aes(label = paste(stat(adj.rr.label), stat(p.value.label), sep = "*\", \"*")),
               formula = y ~ x,
               rr.digits = 3, p.digits = 3,
               parse = TRUE,
               label.x = "left", label.y = "top") +
  labs(title = "KD-37", x = "nsites rep2", y = "nsites rep3") +
  theme_classic() +
  theme(aspect.ratio = 1)


library(gridExtra)

# combine the plots
KD37_nsites_scattar <- grid.arrange(p1, p2, p3, nrow = 1)

ggsave("output/Step06.KD37_nsites_CDS_3UTR_scattar.pdf", plot = KD37_nsites_scattar, device = "pdf", width = 8, height = 6)
```













### Quartile_BoxPlot 
#### SCR, shRNA33, shRNA37, nsites bins cut by SCR nsites value

```{r}

gene_nsites_bin <- gene_nsites %>%  
  rename_all(~ gsub("_nsites", "", .)) %>%
  mutate_all(~ ifelse(is.na(.), 0, .)) %>%
  arrange(desc((SCR_Rep1 + SCR_Rep2 + SCR_Rep3)/3))%>% 
  left_join(Geneid_Genename, by = "Geneid") %>%
  relocate(Genename, .after = Geneid)

# Calculate the number of rows in each bin 
rows_per_bin <- nrow(gene_nsites_bin) / 4

# Print the number of genes in each quartile
print(paste("Each quartile genes number:", rows_per_bin))


# Add a new "bin" column based on the row index
gene_nsites_bin <- gene_nsites_bin %>%
  mutate(bin = as.integer((row_number() - 1) / rows_per_bin) + 1)

# Reshape the data to long format
long_data <- gene_nsites_bin %>%
  gather(key = "sample_Rep", value = "nsites", SCR_Rep1, SCR_Rep2, SCR_Rep3, shRNA33_Rep1, shRNA33_Rep2, shRNA33_Rep3, shRNA37_Rep1, shRNA37_Rep2, shRNA37_Rep3) %>%
  mutate(sample_group = case_when(
    grepl("SCR", sample_Rep) ~ "SCR",
    grepl("shRNA33", sample_Rep) ~ "shRNA33",
    grepl("shRNA37", sample_Rep) ~ "shRNA37"
  ))

# Perform pairwise Wilcoxon rank-sum tests
pairwise_comparisons <- long_data %>%
  group_by(bin) %>%
  pairwise_wilcox_test(nsites ~ sample_group, p.adjust.method = "bonferroni")

# Filter the significant comparisons (e.g., p-value < 0.05)
nsites_3_significant_comparisons <- pairwise_comparisons %>%
  filter(p.adj < 0.05)

print(nsites_3_significant_comparisons)

# Calculate the average EPKM value for each sample group in each bin
average_nsites <- long_data %>%
  group_by(bin, sample_group) %>%
  summarize(mean_nsites = mean(nsites),
            median_nsites = median(nsites),
            .groups = "drop")

# Print the average EPKM values
print(average_nsites)


# save as excel
library(openxlsx)

# Create a new workbook
wb <- createWorkbook()

# Add the data frames as sheets to the workbook
## sheet1
addWorksheet(wb, "gene_nsites")
writeData(wb, "gene_nsites", gene_nsites)

## sheet2
addWorksheet(wb, "nsites_bins")
writeData(wb, "nsites_bins", long_data)

## sheet3
addWorksheet(wb, "3groups_significant")
writeData(wb, "3groups_significant", nsites_3_significant_comparisons)

## sheet4
addWorksheet(wb, "average_nsites")
writeData(wb, "average_nsites", average_nsites)

# Save the workbook to a file
saveWorkbook(wb, "output/Step06.nsites_CDS_3UTR_Quartile_binBynsites_3groups.xlsx", overwrite = TRUE)





# Generate the box plot
colors <- c("#C1C0C0", "#76D7FF", "#1796FF")


nsites_Quartile_binBynsites_3groups <- ggplot(long_data, aes(x = as.factor(bin), y = log2(nsites + 1), fill = sample_group)) +
  geom_boxplot(color = "grey50", outlier.size = 0.5) +
  labs(x = "Quartile", y = "log2(nsites + 1)", title = "nsites (CDS+3'UTR) rank by control nsites value") +
  scale_fill_manual(values = colors, 
                    labels = c("control", "KD-33", "KD-37"),
                    name = "") +
  theme_classic() +
  theme(legend.position = c(0.82, 0.95),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.background = element_rect(fill = "transparent", color = NA),
        legend.text = element_text(size = 9),
        legend.title = element_text(size = 9),
        legend.key.size = unit(0.3, "cm"),
        aspect.ratio = 2/1)



ggsave("output/Step06.nsites_CDS_3UTR_Quartile_binBynsites_3groups.pdf", plot = nsites_Quartile_binBynsites_3groups, device = "pdf", width = 8, height = 6)






```

#### SCR, shRNA, nsites bins cut by SCR nsites value


```{r}
# Reshape the data to long format
long_data <- gene_nsites_bin %>%
  gather(key = "sample_Rep", value = "nsites", SCR_Rep1, SCR_Rep2, SCR_Rep3, shRNA33_Rep1, shRNA33_Rep2, shRNA33_Rep3, shRNA37_Rep1, shRNA37_Rep2, shRNA37_Rep3) %>%
  mutate(sample_group = case_when(
    grepl("SCR", sample_Rep) ~ "SCR",
    grepl("shRNA33", sample_Rep) ~ "shRNA33",
    grepl("shRNA37", sample_Rep) ~ "shRNA37"
  )) %>%
  mutate(combined_group = case_when(
    sample_group == "SCR" ~ "SCR",
    sample_group %in% c("shRNA33", "shRNA37") ~ "shRNA"
  ))

# Perform pairwise Wilcoxon rank-sum tests
pairwise_comparisons <- long_data %>%
  group_by(bin) %>%
  pairwise_wilcox_test(nsites ~ combined_group, p.adjust.method = "bonferroni", pool.sd = FALSE)

# Filter the significant comparisons (e.g., p-value < 0.05)
nsites_2_significant_comparisons <- pairwise_comparisons %>%
  filter(p.adj < 0.05)

print(nsites_2_significant_comparisons)



# Calculate the average EPKM value for each sample group in each bin
average_nsites <- long_data %>%
  group_by(bin, combined_group) %>%
  summarize(mean_nsites = mean(nsites),
            median_nsites = median(nsites),
            .groups = "drop")

# Print the average EPKM values
print(average_nsites)


# save as excel
library(openxlsx)

# Create a new workbook
wb <- createWorkbook()

# Add the data frames as sheets to the workbook
## sheet1
addWorksheet(wb, "gene_nsites")
writeData(wb, "gene_nsites", gene_nsites)

## sheet2
addWorksheet(wb, "nsites_bins")
writeData(wb, "nsites_bins", long_data)

## sheet3
addWorksheet(wb, "2groups_significant")
writeData(wb, "2groups_significant", nsites_2_significant_comparisons)

## sheet4
addWorksheet(wb, "average_nsites")
writeData(wb, "average_nsites", average_nsites)

# Save the workbook to a file
saveWorkbook(wb, "output/Step06.nsites_CDS_3UTR_Quartile_binBynsites_2groups.xlsx", overwrite = TRUE)





# Generate the box plot
colors <- c("#C1C0C0", "#1796FF")


nsites_Quartile_binBynsites_2groups <- ggplot(long_data, aes(x = as.factor(bin), y = log2(nsites + 1), fill = combined_group)) +
  geom_boxplot(color = "grey50", outlier.size = 0.5) +
  labs(x = "Quartile", y = "log2(nsites + 1)", title = "nsites (CDS+3'UTR) rank by control nsites value") +
  scale_fill_manual(values = colors, 
                    labels = c("control", "KD"),
                    name = "") +
  theme_classic() +
  theme(legend.position = c(0.82, 0.95),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.background = element_rect(fill = "transparent", color = NA),
        legend.text = element_text(size = 9),
        legend.title = element_text(size = 9),
        legend.key.size = unit(0.3, "cm"),
        aspect.ratio = 2/1)


ggsave("output/Step06.nsites_CDS_3UTR_Quartile_binBynsites_2groups.pdf", plot = nsites_Quartile_binBynsites_2groups, device = "pdf", width = 8, height = 6)



save(gene_nsites_bin, file = "output/Step06.nsites_CDS_3UTR_Quartile_binBynsites_2groups.rdata")

```




### Differential results
```{r}

gene_nsites <- gene_nsites %>%  
  rename_all(~ gsub("_nsites", "", .)) %>%
  mutate_all(~ ifelse(is.na(.), 0, .)) %>%
  column_to_rownames(var = "Geneid")



# Load necessary libraries
library(limma)
library(edgeR)
library(ggplot2)

# Subset the gene_nsites data frame to include only rows where the row sum is greater than 0
gene_nsites <- gene_nsites[which(rowSums(gene_nsites) > 0),]

# Display the dimensions of the resulting data frame
dim(gene_nsites)

# Create a factor variable condition with three levels
condition <- factor(paste(rep(c("SCR", "shRNA33", "shRNA37"), each = 3)))

coldata <- data.frame(row.names = colnames(gene_nsites), condition)

# Create a DGEList object from the normalized gene expression data
dge <- DGEList(counts = gene_nsites)

# Apply the TMM normalization method using edgeR's calcNormFactors function
dge <- calcNormFactors(dge)

# Convert the normalized data to log2-counts per million (logCPM) using limma's voom function
v <- voom(dge, normalize.method = "none")

# Set up the design matrix for your experimental conditions
design <- model.matrix(~0 + condition)
colnames(design) <- levels(condition)

# Fit the linear model and perform differential expression analysis
fit <- lmFit(v, design)
contrast.matrix <- makeContrasts(shRNA33_vs_SCR = shRNA33 - SCR, shRNA37_vs_SCR = shRNA37 - SCR, shRNA37_vs_shRNA33 = shRNA37 - shRNA33, levels = design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)

# Extract the top differentially expressed genes for each contrast
res_shRNA33_SCR <- topTable(fit2, number = Inf, coef = "shRNA33_vs_SCR")
res_shRNA37_SCR <- topTable(fit2, number = Inf, coef = "shRNA37_vs_SCR")
res_shRNA37_shRNA33 <- topTable(fit2, number = Inf, coef = "shRNA37_vs_shRNA33")

# Perform PCA on the logCPM values
pca <- prcomp(t(v$E), scale = TRUE)

# Create a data frame with the first two principal components and the sample conditions
pca_df <- data.frame(Sample = rownames(pca$x),
                     PC1 = pca$x[, 1],
                     PC2 = pca$x[, 2],
                     Condition = coldata$condition)

# Plot the PCA with ggplot2
ggplot(pca_df, aes(x = PC1, y = PC2, color = Condition)) +
  geom_point(size = 3) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  labs(title = "PCA plot", x = "PC1", y = "PC2") +
  scale_color_discrete(name = "Condition")








# Extract the replicate information from the sample names
pca_df$Replicate <- gsub(".*(Rep\\d+).*", "\\1", pca_df$Sample)

# Plot the PCA with ggplot2 and add sample labels
ggplot(pca_df, aes(x = PC1, y = PC2, color = Condition)) +
  geom_point(size = 3) +
  geom_text(aes(label = Replicate), hjust = 0.3, vjust = 1.3, size = 3) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  labs(title = "PCA plot", x = "PC1", y = "PC2") +
  scale_color_discrete(name = "Condition")




















# create a factor variable condition with three levels
condition2 <- factor(c("SCR", "SCR", "SCR", "shRNA", "shRNA", "shRNA", "shRNA", "shRNA", "shRNA"))


coldata <- data.frame(row.names = colnames(gene_nsites), condition2)

# Create a DGEList object from the normalized gene expression data
dge <- DGEList(counts = gene_nsites)

# Apply the TMM normalization method using edgeR's calcNormFactors function
dge <- calcNormFactors(dge)

# Convert the normalized data to log2-counts per million (logCPM) using limma's voom function
v <- voom(dge, normalize.method = "none")

# Set up the design matrix for your experimental condition2s
design <- model.matrix(~0 + condition2)
colnames(design) <- levels(condition2)

# Fit the linear model and perform differential expression analysis
fit <- lmFit(v, design)
contrast.matrix <- makeContrasts(shRNA_vs_SCR = shRNA - SCR, levels = design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)

# Extract the top differentially expressed genes for each contrast
res_shRNA_SCR <- topTable(fit2, number = Inf, coef = "shRNA_vs_SCR")


# Perform PCA on the logCPM values
pca <- prcomp(t(v$E), scale = TRUE)

# Create a data frame with the first two principal components and the sample condition2s
pca_df <- data.frame(Sample = rownames(pca$x),
                     PC1 = pca$x[, 1],
                     PC2 = pca$x[, 2],
                     condition2 = coldata$condition2)

# Plot the PCA with ggplot2
ggplot(pca_df, aes(x = PC1, y = PC2, color = condition2)) +
  geom_point(size = 3) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  labs(title = "PCA plot", x = "PC1", y = "PC2") +
  scale_color_discrete(name = "condition2")













#### results#############
library(tidyverse)
# res_shRNA33_SCR
res_shRNA33_SCR <- as.data.frame(res_shRNA33_SCR)
res_shRNA33_SCR <- res_shRNA33_SCR %>% 
  rownames_to_column(var = "Geneid") %>%
  left_join(Geneid_Genename, by="Geneid") %>%
  dplyr::select(Geneid, Genename, log2FoldChange = logFC, pvalue = P.Value, padj = adj.P.Val) %>% 
  mutate(direction = if_else(
    padj > 0.05, 'ns', if_else(
      abs(log2FoldChange) == 0, 'ns', if_else(
        log2FoldChange > 0, 'up', 'down')
    ))) %>%  
  arrange(desc(abs(log2FoldChange))) 

## group count
group_by(res_shRNA33_SCR, direction) %>%
  summarise(count=n())


# res_shRNA37_SCR
res_shRNA37_SCR <- as.data.frame(res_shRNA37_SCR)
res_shRNA37_SCR <- res_shRNA37_SCR %>% 
  rownames_to_column(var = "Geneid") %>%
  left_join(Geneid_Genename, by="Geneid") %>%
  dplyr::select(Geneid, Genename, log2FoldChange = logFC, pvalue = P.Value, padj = adj.P.Val) %>% 
  mutate(direction = if_else(
    padj > 0.05, 'ns', if_else(
      abs(log2FoldChange) == 0, 'ns', if_else(
        log2FoldChange > 0, 'up', 'down')
    ))) %>%  
  arrange(desc(abs(log2FoldChange)))   

## group count
group_by(res_shRNA37_SCR, direction) %>%
  summarise(count=n())


# res_shRNA37_shRNA33
res_shRNA37_shRNA33 <- as.data.frame(res_shRNA37_shRNA33)
res_shRNA37_shRNA33 <- res_shRNA37_shRNA33 %>% 
  rownames_to_column(var = "Geneid") %>%
  left_join(Geneid_Genename, by="Geneid") %>%
  dplyr::select(Geneid, Genename, log2FoldChange = logFC, pvalue = P.Value, padj = adj.P.Val) %>% 
  mutate(direction = if_else(
    padj > 0.05, 'ns', if_else(
      abs(log2FoldChange) == 0, 'ns', if_else(
        log2FoldChange > 0, 'up', 'down')
    ))) %>%  
  arrange(desc(abs(log2FoldChange))) 

## group count
group_by(res_shRNA37_shRNA33, direction) %>%
  summarise(count=n())


# res_shRNA_SCR
res_shRNA_SCR <- as.data.frame(res_shRNA_SCR)
res_shRNA_SCR <- res_shRNA_SCR %>% 
  rownames_to_column(var = "Geneid") %>%
  left_join(Geneid_Genename, by="Geneid") %>%
  dplyr::select(Geneid, Genename, log2FoldChange = logFC, pvalue = P.Value, padj = adj.P.Val) %>% 
  mutate(direction = if_else(
    padj > 0.05, 'ns', if_else(
      abs(log2FoldChange) == 0, 'ns', if_else(
        log2FoldChange > 0, 'up', 'down')
    ))) %>%  
  arrange(desc(abs(log2FoldChange))) 

## group count
group_by(res_shRNA_SCR, direction) %>%
  summarise(count=n())














######       save
# save as excel
library(openxlsx)

# Create a new workbook
wb <- createWorkbook()

# Add the data frames as sheets to the workbook
## res_shRNA33_SCR
addWorksheet(wb, "res_shRNA33_SCR")
writeData(wb, "res_shRNA33_SCR", res_shRNA33_SCR)

## res_shRNA37_SCR
addWorksheet(wb, "res_shRNA37_SCR")
writeData(wb, "res_shRNA37_SCR", res_shRNA37_SCR)

## res_shRNA37_shRNA33
addWorksheet(wb, "res_shRNA37_shRNA33")
writeData(wb, "res_shRNA37_shRNA33", res_shRNA37_shRNA33)

## res_shRNA_SCR
addWorksheet(wb, "res_shRNA_SCR")
writeData(wb, "res_shRNA_SCR", res_shRNA_SCR)

# Save the workbook to a file
saveWorkbook(wb, "output/Step06.nsites_CDS_3UTR_DifferentialAnalysis.xlsx", overwrite = TRUE)






res_shRNA33_SCR_nsites <- res_shRNA33_SCR  

res_shRNA37_SCR_nsites <- res_shRNA33_SCR 

res_shRNA37_shRNA33_nsites <- res_shRNA33_SCR 

res_shRNA_SCR_nsites <- res_shRNA33_SCR 


save(res_shRNA33_SCR_nsites,res_shRNA33_SCR_nsites,res_shRNA37_SCR_nsites,res_shRNA37_shRNA33_nsites,res_shRNA_SCR_nsites, file = "output/Step06.nsites_CDS_3UTR_DifferentialAnalysis.rdata")


```


